extensions/amp-story/0.1/animation.js

```diff
@@ -0,0 +1,627 @@
+  hasStarted() {
+    return this.isActivityScheduled_(PlaybackActivity.START) ||
+        this.runner_ && dev().assert(this.runner_)
+            .getPlayState() == WebAnimationPlayState.RUNNING;
+  }
+
+  /** Force-finishes all animations. */
+  finish() {
+    if (this.firstFrameApplied_) {
+      this.notifyFinish_();
+    }
+
+    this.resetFirstFrame_();
+
+    this.playback_(PlaybackActivity.FINISH);
+  }
+
+  /**
+   * @param {!../../amp-animation/0.1/web-animations.WebAnimationRunner} runner
+   * @private
+   */
+  finishWhenReady_(runner) {
+    if (runner.getPlayState() == WebAnimationPlayState.RUNNING) {
+      runner.finish();
+      this.runnerReset_ = true;
+    }
+  }
+
+  /** Cancels animation. */
+  cancel() {
+    this.scheduledActivity_ = null;
+    this.scheduledWait_ = null;
+  }
+
+  /**
+   * @param {!PlaybackActivity} activity
+   * @param {!Promise=} opt_wait
+   */
+  playback_(activity, opt_wait) {
+    const wait = opt_wait || null;
+
+    this.scheduledActivity_ = activity;
+    this.scheduledWait_ = wait;
+
+    if (this.runner_) {
+      this.playbackWhenReady_(activity, wait);
+    }
+  }
+
+  /**
+   * Executes playback activity if runner is ready.
+   * @param {!PlaybackActivity} activity
+   * @param {?Promise} wait
+   */
+  playbackWhenReady_(activity, wait) {
+    const runner = dev().assert(
+        this.runner_,
+        'Tried to execute playbackWhenReady_ before runner was resolved.');
+
+    (wait || Promise.resolve()).then(() => {
+      if (!this.isActivityScheduled_(activity)) {
+        return;
+      }
+
+      this.scheduledActivity_ = null;
+      this.scheduledWait_ = null;
+
+      switch (activity) {
+        case PlaybackActivity.START: return this.startWhenReady_(runner);
+        case PlaybackActivity.FINISH: return this.finishWhenReady_(runner);
+      }
+    });
+  }
+
+  /**
+   * Marks runner as ready and executes playback activity if needed.
+   * @param {!../../amp-animation/0.1/web-animations.WebAnimationRunner} runner
+   * @return {boolean} True if modifies runner state
+   * @private
+   */
+  onRunnerReady_(runner) {
+    this.runner_ = runner;
+
+    runner.onPlayStateChanged(state => {
+      if (state == WebAnimationPlayState.FINISHED) {
+        this.notifyFinish_();
+      }
+    });
+
+    if (!this.isActivityScheduled_()) {
+      return;
+    }
+
+    this.playbackWhenReady_(this.scheduledActivity_, this.scheduledWait_);
+  }
+
+  /**
+   * @param {!PlaybackActivity=} opt_activity
+   * @return {boolean}
+   */
+  isActivityScheduled_(opt_activity) {
+    if (!opt_activity) {
+      return this.scheduledActivity_ !== null;
+    }
+    return this.scheduledActivity_ === opt_activity;
+  }
+
+  /** @private */
+  notifyFinish_() {
+    if (this.target_.id) {
+      this.sequence_.notifyFinish(this.target_.id);
+    }
+  }
+}
+
+
+// TODO(alanorozco): Looping animations
+/** Manager for animations in story pages. */
+export class AnimationManager {
```

Comment:
Minor: This file is large enough to split into two IMO.

Reply:
Will follow-up